docs/developer/api_guidelines.md

```diff
@@ -0,0 +1,438 @@
+# Introduction
+This document serves as a guideline for adding endpoints to the Keptn API,
+and should help to provide a consistent user experience across all of our APIs.
+
+# API Grouping
+We define three different *types* of APIs:
+
+* `public` - customer facing public API exposed by the API gateway
+* `internal` - internally APIs used by internal services never exposed by the API gateway
+* `operations` - internal API providing non-functional features like statistics, debug and testing never exposed by the API gateway
+
+## Public APIs
+Public APIs are always provided by the API gateway and mapped to the K8s Service as follows:
+
+API Gateway: `http://cluster.keptn.sh/api/<component>/<version>/<api resources>`
+
+K8S Service: `<component>/public/<version>/<api resources>`
+
+## Internal APIs
+Internal APIs are not exposed by the API gateway are accessed by other services in Keptn like follows:
+
+K8S Service:  `<component>/internal/<version>/<api resources>`
+
+## Operations APIs
+Operations APIs are not exposed by the API gateway and **should** cover non-functional features
+like statistics, debug or testing APIs etc...
+
+K8S Service: `<component>/operations/<version>/<api resources>` 
+
+---
+
+# Authentication
+Each request **must** have a *Authorization* header containing a base64 encoded bearer string:
+
+## Example
+
+```
+GET /controlplane/v1/sequences
+Authorization: Bearer dGhlc2VjcmV0dG9rZW5mb3JrZXB0bmFwaQ==
+```
+
+Internal communication does not require a token.
+
+---
+
+# Custom Methods
+
+Sometimes, functionality cannot be expressed using the default methods on classical REST resources.
+For this use cases *custom methods* **should** be used:
+
+```
+https://domain/api/<component>/<version>/<api resources>:<custom-method>
+```
+
+- Custom methods **should** use the `HTTP POST` method.
+- The URL **must** end with a verb separated with a colon.
+
+## Example
+```
+POST cluster.keptn.sh/api/controlplane/v1/sequences/94dbe764-e186-48bf-b90d-a53b21fca847:pause
+POST cluster.keptn.sh/api/tokens/v1/token:rotate
+```
+
+ ---
+# Filtering
+
+API endpoints supporting result filtering **must** accept the filter
+as a query parameter named *filter* containing a *filter expression*:
+
+In general each field **should** be filter-able. Exceptions **must** be documented.
+*Filter expressions* follow the following form: `<fieldname> operator <value>`
```

Comment:
I can only filter by a single value. Is this intended? 
Or are filter expressions a serialized array (as of [OpenAPI 3.0](https://swagger.io/docs/specification/serialization/))? If so, will they be considered as `&&` or `||`?

Reply:
Actually, a filter expression is a set of these expressions above combined with boolean operators. I need to describe this better.
However, I begin having the feeling that introducing these kind of filtering is a bit of overkill and actually not needed for the APIs of Keptn. Maybe we are fine with the already existing approach of just using regular query parameter that require a perfect match to the field we want to filter on (basically as we have it right now). Advanced filtering using a dedicated "filter" query parameter with a special simple query language like this could still be considered in the future...
what dou you say?

 @thisthat @bacherfl